### PR TITLE
[RW-2460][risk=no] set up 2FA module backend & cron job to sync from google

### PR DESCRIPTION
**Pull request by jacmrob**

### CODE REVIEW

File: api/src/test/java/org/pmiops/workbench/db/dao/UserServiceTest.java [hunk 1]
```diff
@@ -167,4 +178,13 @@ public void testClearsEraCommonsStatus() throws Exception {
     assertThat(user.getEraCommonsCompletionTime()).isNull();
   }
 
+  @Test
+  public void testSyncTwoFactorAuthStatus() throws Exception {
+    when(directoryService.getUser(EMAIL_ADDRESS)).thenReturn(googleUser);
```
Comment (ericsong) [inline api/src/test/java/org/pmiops/workbench/db/dao/UserServiceTest.java]: it looks like `googleUser` is only used in this test. Should we move narrow down the scope to this function since it is the only one that needs it?

File: api/src/test/java/org/pmiops/workbench/db/dao/UserServiceTest.java [hunk 2]
```diff
@@ -167,4 +178,13 @@ public void testClearsEraCommonsStatus() throws Exception {
     assertThat(user.getEraCommonsCompletionTime()).isNull();
   }
 
+  @Test
+  public void testSyncTwoFactorAuthStatus() throws Exception {
```
Comment (ericsong) [inline api/src/test/java/org/pmiops/workbench/db/dao/UserServiceTest.java]: i think the other case warrants a test case too. should be a quick add.

(setting 2fa completion time to false)

File: api/src/main/java/org/pmiops/workbench/api/ProfileController.java [hunk 3]
```diff
@@ -264,15 +264,17 @@ private String createFirecloudBillingProject(User user) {
...
     if (userAuthentication.getUserType() == UserType.SERVICE_ACCOUNT) {
       // Service accounts don't need further initialization.
       return user;
     }
 
-    Boolean twoFactorEnabled = Optional.ofNullable(user.getTwoFactorEnabled()).orElse(false);
-    if (!twoFactorEnabled) {
-      user.setTwoFactorEnabled(directoryService.getUser(user.getEmail()).getIsEnrolledIn2Sv());
-      user = saveUserWithConflictHandling(user);
+    if (user.getTwoFactorAuthCompletionTime() == null) {
```
Comment (ericsong) [inline api/src/main/java/org/pmiops/workbench/api/ProfileController.java]: nit: we can make this slightly more readable by adding a `completedTwoFactorAuth()` function to the User class which does this check

File: api/src/main/java/org/pmiops/workbench/db/dao/UserService.java [hunk 4]
```diff
@@ -485,4 +489,24 @@ public void syncEraCommonsStatusUsingImpersonation(User user) throws IOException
       }
     }
   }
+
+  /**
+   * Syncs the 2FA status for the current user.
+   */
+  public void syncTwoFactorAuthStatus() {
+    this.syncTwoFactorAuthStatus(userProvider.get());
```
[the remark this answers is not in the document]
Comment (ericsong) [reply api/src/main/java/org/pmiops/workbench/db/dao/UserService.java]: optional: I would say just exposing the parameterized function is sufficient for the UserService's API. The reason why I prefer that is because I think Services (or any code really) should be as stateless as possible so it should try not to import and use stateful dependencies like Provider<User>. Instead, any state related to the request should be captured in the Controllers (which are actually aware of how to interact with HTTP) and passed to the services.

That said, I don't think this is a pattern we have adopted so it's up to you if you want to make the change. Not a big deal either way in this case.

[the remark this answers is not in the document]
Comment (jacmrob) [reply api/src/main/java/org/pmiops/workbench/db/dao/UserService.java]: i'm gonna leave it as-is because we do use this convention in other places and there is a case for when this function is called by the current logged-in user

File: api/src/main/java/org/pmiops/workbench/api/ProfileController.java (hunk 3 above)
Comment (jacmrob) [reply api/src/main/java/org/pmiops/workbench/api/ProfileController.java]: i'm going to argue that that makes this less readable because i have to go click around more, and also adds more boilerplate code

Comment (ericsong) [reply api/src/main/java/org/pmiops/workbench/api/ProfileController.java]: That's fine. But in regards to clicking around, the goal for me is that I shouldn't have to. I don't really care how its implemented, I just want to know if its enabled or not which I should be able to glean from the function name. I had to read through about half the PR before I understood the purpose of the 2fa completion time, and then was able to come back here to reread it and understand whats going on.

File: api/src/main/java/org/pmiops/workbench/db/dao/UserService.java [hunk 5]
```diff
@@ -485,4 +489,23 @@ public void syncEraCommonsStatusUsingImpersonation(User user) throws IOException
...
+
+  public void syncTwoFactorAuthStatus() {
+    syncTwoFactorAuthStatus(userProvider.get());
+  }
+
+  public void syncTwoFactorAuthStatus(User targetUser) {
+    updateUserWithRetries(user -> {
+      Timestamp now = new Timestamp(clock.instant().toEpochMilli());
+      if (directoryService.getUser(user.getEmail()).getIsEnrolledIn2Sv()) {
+        user.setTwoFactorAuthCompletionTime(now);
```
Comment (ericsong) [inline api/src/main/java/org/pmiops/workbench/db/dao/UserService.java]: I believe this will overwrite the twoFactorAuthCompletionTime if it called on a user that already has it set. Is that what we want? This is different from what is happening in ProfileController. ProfileController only sets it if the value is originally null.

Side note: this is another reason for sharing the code here and in ProfileController. Less likelihood of differing implementation and bugs.

Comment (jacmrob) [reply api/src/main/java/org/pmiops/workbench/db/dao/UserService.java]: good catch. will do.

File: api/src/main/java/org/pmiops/workbench/db/dao/UserService.java [hunk 6]
```diff
@@ -485,4 +489,23 @@ public void syncEraCommonsStatusUsingImpersonation(User user) throws IOException
...
     }
   }
+
+  public void syncTwoFactorAuthStatus() {
+    syncTwoFactorAuthStatus(userProvider.get());
+  }
+
+  public void syncTwoFactorAuthStatus(User targetUser) {
+    updateUserWithRetries(user -> {
+      if (user.getTwoFactorAuthCompletionTime() == null &&
```
Comment (ericsong) [inline api/src/main/java/org/pmiops/workbench/db/dao/UserService.java]: i think this actually unsets the user's 2fa completion time if it set to anything. `user.getTwoFactorAuthCompletionTime() = something` -> `user.getTwoFactorAuthCompletionTime() == null` is false -> flows into else statement.

might be good to write a test case to cover this as well as the previous condition. its not a lot of code its covering but boolean statements like this can be tricky so I think its good to have the edge cases covered.

Comment (jacmrob) [reply api/src/main/java/org/pmiops/workbench/db/dao/UserService.java]: thanks for checking.... so here's what i think the logic should be -- since we need to be able to unset the var on our side if the user disables 2FA on the google side (and this method gets called by our cron endpoint that checks that daily to keep them in sync), it should check if the user is enrolled in google, then if the value is null in workbench, set it to a new timestamp.  otherwise (if user is unenrolled in google), set to null in workbench.
